ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,417 @@
+
+            pos_control->set_alt_target(_des_z);
+
+            pos_control->update_z_controller();
+
+            break;
+        }
+
+        case BAIL_OUT:
+        {
+        if (_flags.bail_out_initial == 1) {
+                //functions and settings to be done once are done here.
+
+                #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+                    gcs().send_text(MAV_SEVERITY_INFO, "Bailing Out of Autorotation");
+                #endif
+
+                //Set bail out timer remaining equal to the paramter value
+                if (_param_bail_time < BAILOUT_RAMP_TIME) {
+                    bail_time_remain = BAILOUT_RAMP_TIME;
+                } else {
+                    bail_time_remain = _param_bail_time;
+                }
+
+                //set initial target vertical speed
+                _desired_v_z = curr_vel_z;
+
+                //Initialise position and desired velocity
+                if (!pos_control->is_active_z()) {
+                    pos_control->relax_alt_hold_controllers(arot->get_last_collective());
+                }
+
+                //Get pilot parameter limits
+                float pilot_spd_dn = -get_pilot_speed_dn();
+                float pilot_spd_up = g.pilot_speed_up;
+
+                //set speed limit
+                pos_control->set_max_speed_z(curr_vel_z, pilot_spd_up);
+
+                float pilot_des_v_z = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+                pilot_des_v_z = constrain_float(pilot_des_v_z, pilot_spd_dn, pilot_spd_up);
+
+                //calculate target climb rate adjustment to transition from bail out decent speed to requested climb rate on stick.
+                _target_climb_rate_adjust = (curr_vel_z - pilot_des_v_z)/(_param_bail_time - BAILOUT_RAMP_TIME); //accounting for 0.5s motor spool time
+
+                //calculate pitch target adjustment rate to return to level
+                _target_pitch_adjust = _pitch_target/(_param_bail_time - BAILOUT_RAMP_TIME); //accounting for 0.5s motor spool time_param_bail_time;
+
+                //set acceleration limit
+                pos_control->set_max_accel_z(abs(_target_climb_rate_adjust));
+
+                motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+
+                _flags.bail_out_initial = 0;
+            }
+
+        if (bail_time_remain >= BAILOUT_RAMP_TIME) {
+            //update desired vertical speed and pitch target after the bailout ramp timer has completed
+            _desired_v_z -= _target_climb_rate_adjust*G_Dt;
+            _pitch_target -= _target_pitch_adjust*G_Dt;
+        }
+        //set position controller
+        pos_control->set_alt_target_from_climb_rate(_desired_v_z, G_Dt, false);
+
+        //update controllers
+        pos_control->update_z_controller();
+
+        //update bail out timer
+        bail_time_remain -= G_Dt;
```

Comment:
Think I mentioned it already but normally we would first set a "bail_time_start_ms = AP_HAL::millis()" and then do a check later to see if this is more than XX seconds past now.

Reply:
Done as recommended